Question sharing MVP implementation (#6770)
PrairieLearn/PrairieLearn: docs/questionSharing.md

> # Question Sharing

Author: done

> **NOTE** Question sharing is currently in beta. If your course is hosted on the official PrairieLearn server, you may request that question sharing be turned on for your course. If you administer your own PrairieLearn server, you should _not_ use question sharing at this time. In the future, question sharing will be supported across PrairieLearn servers, so turning on question sharing on unofficial servers at this point in time may lead to naming conflicts in the future.

Reviewer: ```suggestion
**Question sharing is currently in beta.** If your course is hosted on the official PrairieLearn server, you may request that question sharing be turned on for your course. If you administer your own PrairieLearn server, you should _not_ use question sharing at this time. In the future, question sharing may be supported across PrairieLearn servers, so turning on question sharing on unofficial servers at this point in time may lead to naming conflicts in the future.
```

Author: done.

# Question Sharing

In order to avoid instructors needing to copy question files in between courses, PrairieLearn provides a way for questions from one course to be used in assessments in other courses.

## Sharing Names

> In order for another course to import questions from your course into their assessments, you must have chosen a _sharing name_ for your course that they will use as a prefix to your question IDs when importing them. This sharing name will be unique across all PrairieLearn instances and because it will be used in the JSON files for other courses, there will be no way to change the sharing name for your course once you have chosen it. It is recommended that you choose something short but descriptive. For example, if you're teaching a calculus course at a university that goes by the abbreviation 'XYZ', then you could choose the sharing name 'xyz-calculus'. Then other courses will import questions from your course with the syntax `@xyz-calculus/qid`.

Reviewer: I'm just slightly wary of using the term "import" here. Maybe stick with "use", "consume", and similar words?

Author: I've switched it to 'use' for now.

## Sharing a sharing set with connection with another course

For security reasons, establishing the connection for one course to share questions with another course requires coordinated action by owners of the course sharing the questions (which we will refer to hereafter as the 'sharing course') and the course that is using the questions that are being shared ('which we will refer to hereafter as the 'consuming course')

In order to allow someone to share their questions with your course, you must provide them with the 'Sharing ID' listed on the 'Sharing' tab of your instructor settings page. Then the sharing course must use the sharing ID which you provide to them to add your course as a consumer of one of their sharing sets.

## Importing shared questions

> To refer to a question from another course, you simply put the question id (qid) into your `assessmentInfo.json`, prefixed by the `@` symbol and which course it comes from. For example, to use the question `addNumbers` from your own course, you would simply put `"addNumbers"`, but to use the question `addNumbers` from the course with sharing name `test-course`, you would put `@test-course/addNumbers` into your `assessmentInfo.json`. In the context of the `assessmentInfo.json`, this may look like:

Reviewer: A few minor edits:

- Let's avoid the use of words like "simply" in documentation
- Since these docs are describing how to use shared questions, I remove the example demonstrating how to use a question that isn't shared; it felt out of place, especially since it came *before* the example that we're actually writing documentation for.

```suggestion
To refer to a question from another course, use the question id (qid) prefixed by the `@` symbol and the sharing name of the other course. For example, to use the question `addNumbers` from the course with sharing name `test-course`, you would put `@test-course/addNumbers` into your `assessmentInfo.json`. In the context of the `assessmentInfo.json`, this may look like:
```

Author: done

```json
"zones": [
    {
        "title": "Question Sharing Example",
        "comment": "These are new questions created for this exam",
        "questions": [
            {"id": "addNumbers", "autoPoints": [10, 5, 3, 1, 0.5, 0.25]},
            {"id": "@test-course/addNumbers", "autoPoints": [10, 9, 7, 5]}
        ]
    },
]
```

## Steps to share a question for the first time

1. On your course admin page, visit the 'sharing' tab
2. Choose a sharing name for your course
3. Create a sharing set
4. Have the instructor or the course you would like to share your question with visit the 'sharing' tab on their course admin page and provide you with their course's sharing token
5. Use the provided sharing token to add the other instructor's course as a consumer of the sharing set you created
6. Visit the question settings page for the question you would like to share, and add it to the sharing set

> 7. The course you have shared the question with may now use it by referencing it in their assessments

Reviewer: Remind me: is there a way for courses to discover which questions are actually shared with them yet? That is, will a course's questions page display questions that have been shared with them?

Author: No, it will not, that's one of the things we said we would worry about later. 

It gets more complicated to decide what to display when courses are going to start doing public sharing with everyone, so there are some UI details to work out. 

My suggested current workaround is, if you are going to share a bunch of questions with another course, just also  add them as a course content previewer in your course, so they can browser the questions.
